---
title: "MCC Exome seq - compare tumor and cell line"
author: "Keegan Korthauer"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
        code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary

Here we take a look at the expression (exome DNA-seq) data from the samples sequenced 
thus far for both MCC patient tumors, their matching patient-derived cell 
lines, and normal PBMCs.

# Workspace setup

Here we'll load the necessary R packages.

```{r, message = FALSE, warning = FALSE}
library(ggplot2)
theme_set(theme_bw())
library(RColorBrewer) 
library(maftools)
library(tidyverse)

mafdir <- "../../PREPROCESS/DNA/mutect2-gatk4-results/"
```

Read in maf files:

```{r}
maf.files <- list.files(path = mafdir, pattern = "*.maf", full.names = TRUE)

maf <- data.table::fread(maf.files[1])
vcNames <- names(table(maf$Variant_Classification))
maf = maftools:::merge_mafs(maf = maf.files, MAFobj = TRUE)
maf_all = maftools:::merge_mafs(maf = maf.files, MAFobj = TRUE, 
  vc_nonSyn = vcNames)
```

We'll also add in the metadata from the de-identified excel file so we can map
the Broad sample IDs back to the original MCC patient ids. Excel file doesn't 
contain one observation per cell, so we'll have to manually enter the information.

```{r}
# 5350 -> 336
# 5351 -> 350
# 5367 -> 277
# 5368 -> 02-314 (simplified to 314)
# 5369 -> 301
# 5370 -> 286
# 5474 -> 367
# 5473 -> 320
# 5475 -> 290
# 5476 -> 282
# 5477 -> 275 
# 5478 -> 358

maf@clinical.data$patient <- ifelse(grepl("5367", maf@clinical.data$Tumor_Sample_Barcode), "277",
                               ifelse(grepl("5368", maf@clinical.data$Tumor_Sample_Barcode), "314",
                                 ifelse(grepl("5369", maf@clinical.data$Tumor_Sample_Barcode), "301",
                                   ifelse(grepl("5473", maf@clinical.data$Tumor_Sample_Barcode), "320",
                                     ifelse(grepl("5474", maf@clinical.data$Tumor_Sample_Barcode), "367",
                                       ifelse(grepl("5350", maf@clinical.data$Tumor_Sample_Barcode), "336",
                                         ifelse(grepl("5351", maf@clinical.data$Tumor_Sample_Barcode), "350",
                               NA)))))))
maf@clinical.data$type <- ifelse(grepl("-CL-|-C-", maf@clinical.data$Tumor_Sample_Barcode), "Cell Line",
                          "Tumor")


viralneg <- c("286", "350", "290", "282", "320")
maf@clinical.data$virus <- "pos"
maf@clinical.data$virus[maf@clinical.data$patient %in% viralneg] <- "neg"

maf@clinical.data
maf_all@clinical.data <- maf@clinical.data

getSampleSummary(maf)
getSampleSummary(maf_all)
```

# Summary plots

Overall:

```{r, maf_summary, fig.path="plots/", dev="png"}
plotmafSummary(maf = maf, rmOutlier = FALSE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE, 
               showBarcodes = TRUE)
```

```{r, titv_summary, fig.path="plots/", dev="png"}
titv.maf <- titv(maf)
titv.maf$raw.counts
```

```{r, signatures, fig.path="plots/", dev="png"}
mat <- trinucleotideMatrix(maf, ref_genome ='BSgenome.Hsapiens.UCSC.hg19',
         prefix ='chr', add = TRUE, useSyn = TRUE)
require(NMF)
sig <- extractSignatures(mat)
plotSignatures(sig)
```

Compare cell lines to tumors:

```{r, enrichment, fig.path="plots/", dev="png"}
enrich <- clinicalEnrichment(maf, clinicalFeature = "type", minMut = 2)
plotEnrichmentResults(enrich, pVal = 0.1)
```


Compare viral pos to viral neg:

```{r, enrichmentvirus, fig.path="plots/", dev="png"}
enrich <- clinicalEnrichment(maf, clinicalFeature = "virus", minMut = 2)
plotEnrichmentResults(enrich, pVal = 0.1)
```

## Tumors only

```{r, maf_summary_T, fig.path="plots/", dev="png"}
tumors <- maf@clinical.data$Tumor_Sample_Barcode
tumors <- tumors[grepl("T", tumors)]
mafT <- subsetMaf(maf, tsb = tumors, mafObj = TRUE)
plotmafSummary(maf = mafT, rmOutlier = FALSE, addStat = 'median', 
               dashboard = TRUE, titvRaw = FALSE, showBarcodes = TRUE)
```

```{r, titv_summary_T, fig.path="plots/", dev="png"}
titv.maf <- titv(mafT)
titv.maf$raw.counts
```

```{r, signatures_T, fig.path="plots/", dev="png", eval = FALSE}
mat <- trinucleotideMatrix(mafT, ref_genome ='BSgenome.Hsapiens.UCSC.hg19',
         prefix ='chr', add = TRUE, useSyn = TRUE)
sig <- extractSignatures(mat)
plotSignatures(sig)
```

## Cell lines only

```{r, maf_summary_CL, fig.path="plots/", dev="png"}
cl <- maf@clinical.data$Tumor_Sample_Barcode
cl <- cl[grepl("CL|-C", cl)]
mafCL <- subsetMaf(maf, tsb = cl, mafObj = TRUE)
plotmafSummary(maf = mafCL, rmOutlier = FALSE, addStat = 'median', 
               dashboard = TRUE, titvRaw = FALSE, showBarcodes = TRUE)
```

```{r, titv_summary_CL, fig.path="plots/", dev="png"}
titv.maf <- titv(mafCL)
titv.maf$raw.counts
```

```{r, signatures_CL, fig.path="plots/", dev="png", eval=FALSE}
mat <- trinucleotideMatrix(mafCL, ref_genome ='BSgenome.Hsapiens.UCSC.hg19',
         prefix ='chr', add = TRUE, useSyn = TRUE)
sig <- extractSignatures(mat)
plotSignatures(sig)
```


## Viral positive only

```{r, maf_summary_Pos, fig.path="plots/", dev="png"}
pos <- maf@clinical.data$Tumor_Sample_Barcode
pos <- pos[maf@clinical.data$type == "pos"]
mafP <- subsetMaf(maf, tsb = tumors, mafObj = TRUE)
plotmafSummary(maf = mafP, rmOutlier = FALSE, addStat = 'median', 
               dashboard = TRUE, titvRaw = FALSE, showBarcodes = TRUE)
```

```{r, titv_summary_Pos, fig.path="plots/", dev="png"}
titv.maf <- titv(mafP)
titv.maf$raw.counts
```

```{r, signatures_Pos, fig.path="plots/", dev="png", eval = FALSE}
mat <- trinucleotideMatrix(mafP, ref_genome ='BSgenome.Hsapiens.UCSC.hg19',
         prefix ='chr', add = TRUE, useSyn = TRUE)
sig <- extractSignatures(mat)
plotSignatures(sig)
```

## Viral negative only

```{r, maf_summary_Neg, fig.path="plots/", dev="png"}
neg <- maf@clinical.data$Tumor_Sample_Barcode
neg <- neg[maf@clinical.data$type == "neg"]
mafN <- subsetMaf(maf, tsb = tumors, mafObj = TRUE)
plotmafSummary(maf = mafN, rmOutlier = FALSE, addStat = 'median', 
               dashboard = TRUE, titvRaw = FALSE, showBarcodes = TRUE)
```

```{r, titv_summary_Neg, fig.path="plots/", dev="png"}
titv.maf <- titv(mafN)
titv.maf$raw.counts
```

```{r, signatures_Neg, fig.path="plots/", dev="png", eval = FALSE}
mat <- trinucleotideMatrix(mafN, ref_genome ='BSgenome.Hsapiens.UCSC.hg19',
         prefix ='chr', add = TRUE, useSyn = TRUE)
sig <- extractSignatures(mat)
plotSignatures(sig)
```

# Rainfall plots

```{r, echo=FALSE}
dev.off()
```

```{r rainfall, fig.height = 5, fig.width = 7, fig.path="plots/", dev="png"}
for (s in maf@clinical.data$Tumor_Sample_Barcode){
  p <- rainfallPlot(maf, tsb = s)
  print(p)
}
```

# Clusters

```{r heterogeneity, fig.path="plots/", dev="png"}
clusters <- inferHeterogeneity(maf)
clusters

for (s in unique(clusters$clusterMeans$Tumor_Sample_Barcode)){
  p <- plotClusters(clusters, tsb = s)
  print(p)
}
```



# Co-mut plots

```{r comut, fig.height = 8.5, fig.width = 6, fig.path="plots/", dev="png"}
fabcolors = RColorBrewer::brewer.pal(n = 7,name = 'Set1')
names(fabcolors) = c("320", "367", "277", "314", "301", "336", "350")
typecols <- rev(c("black", "grey"))
names(typecols) <- c("Cell Line", "Tumor")
virus <- c(pos = "#543005", neg = "#DFC27D")
fabcolors = list(type = typecols,
                 virus = virus,
                 patient = fabcolors)

oncoplot(maf = maf, top = 50,
         clinicalFeatures = c("type", "virus", "patient"),
         annotationColor = fabcolors,
         fontSize = 0.8) #, 
      #  sampleOrder = c("DFCI-5473-C-01", "DFCI-5473-T-01",
      #                   "DFCI-5474-C-01", "DFCI-5474-T-01",
      #                   "DFCI-5367-CL-01", "DFCI-5367-T-01",
      #                   "DFCI-5368-CL-01", "DFCI-5368-T-01",
      #                   "DFCI-5369-CL-01", "DFCI-5369-T-01"))
```

## Co-mut plots for AP genes

```{r comut-goi, fig.height = 8.5, fig.width = 6, fig.path="plots/", dev="png"}
goi <- read_tsv("../genes_of_interest.txt", col_names = FALSE)
colnames(goi) <- c("symbol", "description", "AP")

ap_genes <- goi$symbol[goi$AP & !is.na(goi$AP)]

#pdf("plots/oncoplot_apgenes.pdf", height = 10)
oncoplot(maf = maf, genes=ap_genes, 
         clinicalFeatures = c('type', "patient"),
         annotationColor = fabcolors,
         removeNonMutated = FALSE,
         fill = TRUE, # have to use github version to have this arg)
         fontSize = 0.8)
#dev.off()
```

# clustering

Need to construct gene by sample matrix from the tidy matrix in `maf@data`.

```{r}
tidy_maf <- maf@data

tidy_maf <- select(tidy_maf, sample_id, Hugo_Symbol) %>%
  group_by(sample_id, Hugo_Symbol) %>%
  summarize(muts = n())

mat_maf <- spread(tidy_maf, sample_id, muts, 0) %>%
  column_to_rownames(var = "Hugo_Symbol")

head(mat_maf)
```

Now we're ready to compute distances.

```{r}
d <- dist(t(mat_maf), method = "binary")
plot(hclust(d))
```
# Lollipop plot for ERAP1 mutations

We are interested in characterizing the location and consequence of the mutations in 
ERAP1.

```{r lollipop-erap, fig.height = 4, fig.width = 6, fig.path="plots/", dev="png"}
# list genes with mutation in ERAP1
erap <- subsetMaf(maf = maf_all, genes = c('ERAP1'), mafObj = FALSE)
erap

write.table(erap, quote=FALSE, file = "plots/erap_mutations.txt", sep = "\t", row.names = FALSE)

# plot
pdf("plots/lollipop-erap.pdf", width = 10, height = 3)
lollipopPlot(maf = maf_all, gene = 'ERAP1',  showMutationRate = TRUE)
dev.off()
```
# Session Information

```{r}
sessionInfo()
```
